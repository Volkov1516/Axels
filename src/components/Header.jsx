import { useSelector, useDispatch } from 'react-redux';
import 'bootstrap/dist/css/bootstrap.min.css';

import { Navbar, Container, Button } from 'react-bootstrap';

import { SET_USER } from '../redux/features/auth/authSlice.js';
import { UserName, NavbarBrand, BrandLink } from '../styled/Header.js';

const Header = () => {
  const dispatch = useDispatch();

  const { isAuth } = useSelector(state => state.auth);

  const handleLogout = () => {
    dispatch(SET_USER(false));
  }

  return (
    <Navbar expand="lg" bg="primary">
      <Container>
        <BrandLink to="/">
          <NavbarBrand> Home </NavbarBrand>
        </BrandLink>
        {isAuth && (
          <div className="justify-content-end">
            <UserName>{isAuth.email}</UserName>
            <Button onClick={handleLogout} variant="outline-light" size="sm">
              Logout
            </Button>
          </div>
        )}
      </Container>
    </Navbar>
  )
};

export default Header;
