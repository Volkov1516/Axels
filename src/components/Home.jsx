import 'bootstrap/dist/css/bootstrap.min.css';

import { Button, Container, Row, Col } from 'react-bootstrap';

import { 
  ProductCard, ProductImg, ProductTitle, ProductDescription, ProductImgWrapper, 
  ProductLink
} from '../styled/Home.js';

const Home = ({ products }) => (
  <Container>
    <Row>
      {products &&
        products.map((product, index) => (
          <Col key={index} lg="3" md="4" sm="6" >
            <ProductCard>
              <ProductImgWrapper>
                <ProductImg src={product.img} alt="product img" variant="top" />
              </ProductImgWrapper>
              <ProductTitle> {product.title} </ProductTitle>
              <ProductDescription>{product.description}</ProductDescription>
              <ProductLink to={`/product/${product.id}`}>
                <Button variant="primary">Details</Button>
              </ProductLink>
            </ProductCard>
          </Col>
        ))}
    </Row>
  </Container>
);

export default Home;
