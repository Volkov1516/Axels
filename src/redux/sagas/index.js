import { takeEvery, put, call } from 'redux-saga/effects';

import { GET_PRODUCTS } from '../features/product/productSlice';
import { GET_REVIEWS } from '../features/review/reviewSlice';
import { getData } from '../../api';

export function* workerSaga() {
  const data = yield call(getData)

  yield put({type: GET_PRODUCTS, payload: data.products})
  yield put({type: GET_REVIEWS, payload: data.reviews})
}

export function* watcherSaga() {
  yield takeEvery('LOAD_DATA', workerSaga)
}

export default function* rootSaga() {
  yield watcherSaga()
}