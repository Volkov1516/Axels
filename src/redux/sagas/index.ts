import { takeEvery, put, call, SagaReturnType } from 'redux-saga/effects';

import { GET_PRODUCTS } from '../features/product/productSlice';
import { GET_REVIEWS } from '../features/review/reviewSlice';
import { getData } from '../../api';

type GetData = SagaReturnType<typeof getData>

export function* workerSaga() {
  const data: GetData = yield call(getData);

  yield put({type: GET_PRODUCTS, payload: data.products});
  yield put({type: GET_REVIEWS, payload: data.reviews});
}

export function* watcherSaga() {
  yield takeEvery('LOAD_DATA', workerSaga);
}

export default function* rootSaga() {
  yield watcherSaga();
}

