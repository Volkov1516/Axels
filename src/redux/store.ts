import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from '@redux-saga/core';

import authReducer from './features/auth/authSlice';
import productReducer from './features/product/productSlice';
import reviewReducer from './features/review/reviewSlice';
import rootSaga from './sagas';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    auth: authReducer,
    product: productReducer,
    review: reviewReducer
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(sagaMiddleware)
})

sagaMiddleware.run(rootSaga);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

