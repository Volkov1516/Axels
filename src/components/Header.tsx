import 'bootstrap/dist/css/bootstrap.min.css';

import { Navbar, Container, Button } from 'react-bootstrap';

import { SET_USER } from '../redux/features/auth/authSlice';
import { UserName, NavbarBrand, BrandLink } from '../styled/Header.js';
import { useAppSelector, useAppDispatch } from '../redux/hooks'

const Header = () => {
  const dispatch = useAppDispatch();

  const { isAuth } = useAppSelector(state => state.auth);

  const handleLogout = () => {
    dispatch(SET_USER({
      id: 0,
      email: '',
      password: '',
      logged: false
    }));
  }

  return (
    <Navbar expand="lg" bg="primary">
      <Container>
        <BrandLink to="/">
          <NavbarBrand> Home </NavbarBrand>
        </BrandLink>
        {isAuth.logged && (
          <div className="justify-content-end">
            <UserName>{isAuth.email}</UserName>
            <Button onClick={handleLogout} variant="outline-light" size="sm">
              Logout
            </Button>
          </div>
        )}
      </Container>
    </Navbar>
  )
};

export default Header;
