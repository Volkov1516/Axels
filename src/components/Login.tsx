import React, {useState} from 'react';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import 'bootstrap/dist/css/bootstrap.min.css';

import { Container, Form } from 'react-bootstrap';

import { SET_USER } from '../redux/features/auth/authSlice';
import { FormWrapper, SubmitButton, FormWarning } from '../styled/Login';
import { useAppDispatch } from '../redux/hooks';

const Login = () => {
  const dispatch = useAppDispatch();

  const [toggleSign, setToggleSign] = useState(false);

  const formik = useFormik({
    initialValues: {
      email: '',
      password: '',
    },
    validationSchema: Yup.object({
      email: Yup.string().required('Required!').min(8, 'Too Short!').email('Invalid email!'),
      password: Yup.string().required('Required!').min(8, 'Too Short!').max(20, 'Too long!'),
    }),
    onSubmit: (values) => {
      dispatch(SET_USER({
        id: 1,
        email: values.email,
        password: values.password,
        logged: true
      }))
    },
  });

  const SignUp = () => (
    <FormWrapper>
      <Form onSubmit={formik.handleSubmit}>
        <Form.Group>
          <Form.Label>Email:</Form.Label>
          <Form.Control 
            id="email" name="email" type="text" placeholder="example@mail.com" 
            value={formik.values.email} onChange={formik.handleChange} 
          />
          {formik.errors.email && <FormWarning>{formik.errors.email}</FormWarning>}
        </Form.Group>
        <Form.Group>
          <Form.Label>Password:</Form.Label>
          <Form.Control 
            id="password" name="password" type="password" placeholder="Create a password" 
            value={formik.values.password} onChange={formik.handleChange} 
          />
          {formik.errors.password && <FormWarning>{formik.errors.password}</FormWarning>}
        </Form.Group>
        <SubmitButton type="submit" > Sign Up </SubmitButton>
        <SubmitButton onClick={() => setToggleSign(!toggleSign)} variant="outline-danger" size="sm" >
          Sign In
        </SubmitButton>
      </Form>
    </FormWrapper>
  )

  const SignIn = () => (
    <FormWrapper>
      <Form onSubmit={formik.handleSubmit}>
        <Form.Group>
          <Form.Label>Email:</Form.Label>
          <Form.Control 
            id="email" name="email" type="text" placeholder="example@mail.com" 
            value={formik.values.email} onChange={formik.handleChange} 
          />
          {formik.errors.email && <FormWarning>{formik.errors.email}</FormWarning>}
        </Form.Group>
        <Form.Group>
          <Form.Label>Password:</Form.Label>
          <Form.Control 
            id="password" name="password" type="password" placeholder="Create a password" 
            value={formik.values.password} onChange={formik.handleChange} 
          />
          {formik.errors.password && <FormWarning>{formik.errors.password}</FormWarning>}
        </Form.Group>
        <SubmitButton type="submit" > Sign In </SubmitButton>
        <SubmitButton onClick={() => setToggleSign(!toggleSign)} variant="outline-danger" size="sm" >
          Sign Up
        </SubmitButton>
      </Form>
    </FormWrapper>
  )

  return (
    <Container>
      { toggleSign ? <SignIn /> : <SignUp /> }
    </Container>
  )
}

export default Login;
