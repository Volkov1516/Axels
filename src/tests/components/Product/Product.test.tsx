import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';

import { Product, Comments } from '../../../components/index';

import { store } from '../../../redux/store';

configure({adapter: new Adapter()});

const product = {
  id: 0,
  title: '',
  description: '',
  descriptionFull: '',
  img: ''
};

const review = {
  id: 0,
  productId: 0,
  userId: 0,
  userName: '',
  text: '',
  rate: ''
};

let wrapper: any;

describe('Login', () => {

  beforeAll(() => {
    wrapper = shallow(<Provider store={store} ><Product product={product} ></Product></Provider>);
  });

  it('should take a snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

});

