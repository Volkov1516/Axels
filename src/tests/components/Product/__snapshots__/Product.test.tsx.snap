// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should take a snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Product
    product={
      Object {
        "description": "",
        "descriptionFull": "",
        "id": 0,
        "img": "",
        "title": "",
      }
    }
  />
</Provider>
`;

exports[`Login takes a snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Product
    product={
      Object {
        "description": "",
        "descriptionFull": "",
        "id": 0,
        "img": "",
        "title": "",
      }
    }
  />
</Provider>
`;
